src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -0,0 +1,280 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.Extensions.Metrics;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+
+internal sealed class KestrelMetrics
+{
+    // Note: Dot separated instead of dash.
+    public const string MeterName = "Microsoft.AspNetCore.Server.Kestrel";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentConnectionsCounter;
+    private readonly Histogram<double> _connectionDuration;
+    private readonly Counter<long> _rejectedConnectionsCounter;
+    private readonly UpDownCounter<long> _queuedConnectionsCounter;
+    private readonly UpDownCounter<long> _queuedRequestsCounter;
+    private readonly UpDownCounter<long> _currentUpgradedRequestsCounter;
+    private readonly Histogram<double> _tlsHandshakeDuration;
+    private readonly UpDownCounter<long> _currentTlsHandshakesCounter;
+
+    public KestrelMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentConnectionsCounter = _meter.CreateUpDownCounter<long>(
+           "current-connections",
+            description: "Number of concurrent connections that are currently active on the server.");
+
+        _connectionDuration = _meter.CreateHistogram<double>(
+            "connection-duration",
+            unit: "s",
+            description: "The duration of connections on the server.");
+
+        _rejectedConnectionsCounter = _meter.CreateCounter<long>(
+           "rejected-connections",
+            description: "Number of connections rejected by the server. Connections are rejected when the currently active count exceeds the value configured with MaxConcurrentConnections.");
+
+        _queuedConnectionsCounter = _meter.CreateUpDownCounter<long>(
+           "queued-connections",
+            description: "Number of connections that are currently queued and are waiting to start.");
+
+        _queuedRequestsCounter = _meter.CreateUpDownCounter<long>(
+           "queued-requests",
+            description: "Number of HTTP requests on multiplexed connections (HTTP/2 and HTTP/3) that are currently queued and are waiting to start.");
+
+        _currentUpgradedRequestsCounter = _meter.CreateUpDownCounter<long>(
+           "current-upgraded-requests",
+            description: "Number of HTTP requests that are currently upgraded (WebSockets).");
```

Comment:
Connections are upgraded, not requests.

It doesn't include HTTP/3 WebTransport requests, right?